app/models/docker_builder_service.rb

```diff
@@ -51,11 +50,15 @@ def run!(push: false, tag_as_latest: false)
     @execution = JobExecution.new(build.git_sha, job) do |_, tmp_dir|
       if build.kubernetes_job
         run_build_image_job(job, push: push, tag_as_latest: tag_as_latest)
-      elsif build_image(tmp_dir)
-        ret = true
-        ret = push_image(tag_as_latest: tag_as_latest) if push
-        build.docker_image.remove(force: true) unless ENV["DOCKER_KEEP_BUILT_IMGS"] == "1"
-        ret
+      else
+        if build_image(tmp_dir) # rubocop:disable Style/IfInsideElse
+          ret = true
+          ret = push_image(tag_as_latest: tag_as_latest) if push
+          build.docker_image.remove(force: true) unless ENV["DOCKER_KEEP_BUILT_IMGS"] == "1"
```

Comment:
Are you thinking of continuing to use the docker gem to handle pushing the image and deleting the image?